plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Instance/Flat.hs

```diff
@@ -121,6 +125,19 @@ instance ( Closed uni
          , Flat fun
          , Flat ann
          , Flat name
+         , Flat (Binder name)
          ) => Flat (Program name uni fun ann) where
     encode (Program ann v t) = encode ann <> encode v <> encode t
     decode = Program <$> decode <*> decode <*> decode
+
+-- | We use this overriding newtype (instead of resorting to overlapping instances)
+-- to skip the (de)serializing if the LamAbs Binder is DeBruijn
+newtype Binder a = Binder { unBinder :: a }
+
+instance Flat (Binder DeBruijn) where
+    size _ = id -- zero cost
+    encode _ = mempty
+    decode = pure $ Binder (DeBruijn 0)
+
+deriving newtype instance Flat (Binder Name)
+deriving newtype instance Flat (Binder NamedDeBruijn)
```

Comment:
You could encode a named debruijn as just a name in the binder and just an index as a variable. Not that it really matters, since I don't think we use this anywhere much.